src/dvi/tmds.rs

```diff
@@ -23,33 +23,53 @@ impl TmdsSymbol {
     pub const C3: Self = TmdsSymbol(0x2ab);
 
     pub const fn encode(discrepancy: i32, byte: u32) -> (i32, Self) {
-        let byte_ones = u8::count_ones(byte as u8);
-        let a = (byte << 1) ^ byte;
-        let b = (a << 2) ^ a;
-        let mut c = ((b << 4) ^ b) & 0xff;
+        let byte_ones = u8::count_ones(byte as _);
 
-        if byte_ones > 4 || (byte_ones == 4 && (c & 1) == 0) {
-            c ^= 0xaa;
+        // The first step of encoding TMDS data is to XOR/XNOR each input bit with the previous output bit, one by one
+        //
+        // Instead using a for loop over each bit, TMDS encoding can be approached
+        // as a carry-less multiplication with 255. The decoding step would be a
+        // carry-less multiplication with 3.
+        //
+        // To reduce the amount of steps, the carry-less multiplication with 255
+        // can be split up into a multiplication with 3 * 5 * 17. The following
+        // 3 lines respectively can be seen as those smaller multiplications.
+        let byte_mul = (byte << 1) ^ byte;
+        let byte_mul = (byte_mul << 2) ^ byte_mul;
+        let byte_mul = (byte_mul << 4) ^ byte_mul;
+        // We only care about the bottom byte
+        let mut byte_encoded = byte_mul & 0xff;
+    
+        let should_xnor = byte_ones > 4 || (byte_ones == 4 && (byte_encoded & 1) == 0);
+    
+        if should_xnor {
+            // Convert the XOR case to the XNOR case, toggling every other bit
+            byte_encoded ^= 0xaa
         } else {
-            c ^= 0x100;
-        }
-
-        let mut c_ones = u8::count_ones(c as u8);
-
-        let invert = if discrepancy == 0 || c_ones == 4 {
-            (c >> 8) == 0
+            // Set the high bit to indicate XOR
+            byte_encoded ^= 0x100
+        };
+    
+        let mut byte_encoded_ones = u8::count_ones(byte_encoded as _);
+    
+        let should_invert = if discrepancy == 0 || byte_encoded_ones == 4 {
+            (byte_encoded >> 8) == 0
         } else {
-            (discrepancy > 0) == (c_ones > 4)
+            (discrepancy > 0) == (byte_encoded_ones > 4)
         };
-
-        if invert {
-            c ^= 0x2ff;
-            c_ones = 9 - c_ones;
+    
+        if should_invert {
+            // Invert the lower byte and set the 10th bit
+            byte_encoded ^= 0x2ff;
+            byte_encoded_ones = 9 - byte_encoded_ones;
```

Comment:
After this block, the variable contains the ones count of the lower 8 bits plus bit 9. It's missing bit 8 because that doesn't get flipped by inversion. I'm not sure if that needs to be stated explicitly but it is part of understanding the code.

Reply:
I have moved the things around a bit, and tried to explain what is going on. But I'm not set on this being the best way of writing it.

In making this change (the only one that wasnt comment updates), The amount of branches has decreased by one and the instruction count has gone down a little ([godbolt link](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgAzADZSjgBk8BkwAOXcAI0xiCWkAB1QFQjsGFzcPbz8EpNsBIJDwliiYqUtMa1yGIQImYgI0909fMoqU6tqCfLDI6NjLGrqGjObzDq7C4tiASktUE2Jkdg4AUgAmLzAwZYBWACEGVAB9c3QdgBFljQBBS5v1nd2sYjwAN0wIJ1oBTFIAahc4gBPKbnW5xEwRX7mUw2X4AFRY6AUQkBRVcEBMXlWIK8u1utzwLDitHhiORqIirl%2BywA7Hjrr9Gb9wZC0AxzP8NCBfkJylRqV4zqSkSi0bQIBpVF5tpIcfSrkzmRDfmyOU4uNzebR%2BctBcLyWKJaomBE5bdFSyVQI1atNXyBUKESKKejJVwZWaGUzLaqCP8vHbtQ79aLKeLJasTZ6bl7GT7rX6qAxfow0FgIPgFMhiJg4oZkIDuXgsX8IoCCOxfpjsb8ALS67C/CDF1Z/LVUKbUunmxVM%2Bh%2B0bATCHLiHMsVw7fBTBkwADhAKDmgknIQUEHHmF%2BTGnc%2Bjvb7mD9TGD6/Lm91Tgvvy4nZ2jY3uvl%2B8Z/d%2BkN1Qogx4vV5rd63j49vur4sCYfrICe64CpeXhOL8srUts963qsPi/JKVBUIB1xAb2eD8oOw6jhuK6YNODbwV2l40nBECESOY5nqR5GCh%2BlFrD4HFNhu1Kodet6sXqGi3t2sbPoyEF3mxkpMEw2EKs%2BtJCuUShdk%2B4lMpJSHSaoXAaBo8niUp%2BJiYqIFgSqzEzvOi4mMuU4QBB25VrO0a4WZh6/MEbx1MG%2BG/Jm2a5vmgIOtJVG0nByBWR%2BbGSGp7nPo5ArYBRrlhUJiWMkpKa0KptLqRpTaBTmeYMAWKXoQJFx6o5MVeI2spZV2NXys1/nedEfoFc1ipaalQmqKsmGGUVllTsGACcdbjauo37sZOGmZpMWrHitWSQ1aUoWhXDyc1GZ4FmpUhbxuxNtFE3OS2M3bDMIYuuKyBTCCYmLTcNJnBwMy0Jw2y8J4HBaKQqCcAASmYA5zAs57rDwpAEJo32kBASBpjg%2BBEGQFDUHQjCsEs8P8IIIhiOwpTE/IShqEjpC6K2BhGCApjmC0mA2CkDgMM4riNN40iBME3RFL07qkNkyQCEMTTxIkksMOMPQlNsbMcwI7SDLzwzSFY7OVBrnRCxMosq6MmvpDL/RjEbSsSHdszzIsnBrBsWx7AcxwEKc2wXEtdxeA8TyvO8nzfH8ALAqC1yWtCJiwk6BphhiWJudchLEg9YoJWJ8bsn6Thcjy9psQnoaulKHrzUqrIJv8GpF0GJdkmX4aqBoUZV7nNqBjqeql49Rruk1uK4V3%2BcBg3veOs3A8Rh3I9%2Bxayq%2Br8SYpuVGDDi8qyHcdwXlYWXklm%2BZ7ctWnb1g1TYtm2fIiYVvaviRE1sXOC5oHZBDMaeFZbtOhypwUvuAA9MA%2BECBNw0GIBycwuZfioH5KmDAwRgDwgALJnCEAFJgNQvLTiIL8AAGgAeTBsAwhoRSEpiYMgBAXkGDgj9BEQgvwADuhA6EEAgcyHMLw7BmHgWBRhb5CB/G%2BCfeBIRmqgOkWAgAknnTATB0BViSEYLcq8SC/C%2BKgOI8CfLUNoSIiw6DMHrzTCglUhg3ybiYHEOIxBUA0IgacZajJQF/w0aIYgxBAS1noAoacoFaC2GJHgUQlQ2EcN%2BKsbY2wAB04DNxYAseo2BejWFzFoCoqIAE3G/A8d43x/iyJBJMCEvAYSIkpCiVw34Xh4myNkfCVAvwczoDjpuLhtiWBLj9AgqEFY4gKD%2BN0qxPi/EBLKRUqpOCansLqbE7YzTRDJlyQoMJfoTB6OCAQ48wTQm0HCXMgQtS6FeF%2BAAKl%2BNsK514aSJLhNw/gtAvjsKMM0i5RzVxtLInEPWwdaChVWTYqEmBGCeK4YkTcCgWBiHoMQX4BzKlHOqdaRp%2BSn5MWCSeHiP5YL8UQveM8VcPJMOxeU3FFKST4rgn%2BJCJ8JzBNJQeclTLKVsR/sOHFtL4K3gZSRZlC8gG9g8QAdU3AIIF4zbGUgsmMykBAiAsEZZgZq5k2XDiQVgFRbFBUcr4hhLCwr1WeQUAgLJ6BDiqAOIivVTEX5X3ipFSK3EHU/NinqZ1qEuJcsONqzAuq%2BI3gykKYS%2B18n%2BXNZa61trs4ivEh4lwDAfJ%2BjGSQsGVjVIEPTZQzNoglCjNQMAYA3zUGYB8qFVA3TEXMIIL1JkJEA26u0oNWSzUcoqXPKJBNz4PG8iYSw2cvwCHBHwBEzcGaG2MibRvHVRKdJ6QMvk4yJrMWeWbfVIUb9bL2VXH65tniAERt7S%2BM1FrylWs6r5Ni/kSr7wqp6sNEUaLmM3lax1NUhTOp7WNN1E5D0Nm2qG9CHbPq5Xyr%2Bsau8gplUfVfYSIGICbsdY1V6p6WonvEk/Qghxh2ctnejXVW0r7pS4ntYV2GzWPS3V5AiF7smHGvd1KDRUPEKNTSO7hbzoiqq3NzMFg6/STWnaq/1c7A0LsGsNY1bV8mKnYymrqXHJU/I/oIeB/Ixk8cRcOutIytzoByUO/jxnhOiYgMO2sb6dXMRQudOteGzrXjA8pPK3aH7iRQx6taxi8Ouaw8%2BV8964OhTYiF06axzp0Ro1dacN1rN3UC/uGDJ0D5/H7oaQj76Xq4XekpZGv0OD/VIIDYGoMOAQxgdDRYvEsS8ERkDb6MwADW3gGmSCkEPHwXXJo%2BBlPoTgkgAa0wq7wBQehGtaBmHAWAaNUBEjxtjSgaBFsIuZlwGkelSBYD4YsAAangTArDiH/MBkTOgFZoGUAiLTZhzBfGcHhvd2ogJiERG0HrJ7vBVtsEEMQhgQLaZYAiCYYATh4UTe4LwLAcKjDiCaztvAOYOZvCh8DTAqh2ZgUJrwXZ5RaZHIiMQV7LgsC0wIM8Fg32ZhUAMMABQh3juncYN9mQJNRDiAprIRQKh1CI90FwfQhhjCQ30HgCIE3IAzF0ZUKHtZiG8FQD5Z4WApdQGYGwEAgb5akBeGIEwSxVgaFWDwGbmv2C1FoXrg37AuGRBeD4FgtYDghFrEo5hL1VaVC5jzC2ehBYFFtkLiWlRpZ6FDykRWIsSje7aAMeoWtPBC91mrKoCfo%2BTBTwn8P2frZB5jxIGYCgavkx%2Bn9EbiOKttMhhBLg8StvxI0E2XAhAtEuxvA1pGM20ZUGAOQFbC24hLdCATTgLMCB14b/XjHmMSB4HQHoQX3vA0ADE7Jp84Jj1gxJ2AgEy2GBcAbOAIxzJgBfIBhKDeK5X8rnAnCr4AOI1/MFPxvzeICt6xnVoX/wh9LZ/ymC7yay9wgWUV6AgFa3a3iVWEmkkEmikB8H622BpG2B8FiWv2G1IGpyXVK1G04HG0m27xRhgEQEXDW2iAH0cn/3W2IE222123CUwCZxOzOzZxoBCWiAm3XDu2CFezZxe18Xe0%2BxsDZ1%2B0YAIAByB0RxBzBwh1eSh3hlhxFwR2BnwBR1sDR1pkx2xwrDZ3xyK2BiJxJ18TJyWGBkp0JBpz4Hp0ZyO1YNZ2h3Z2EE53JmkEpl5xpgFxAAZhF2ZjFyJ3VxlziDl04AVyVxVwXzIngAgAt210xhSBtzcCNxNzNxRjiKtwQCSMNxAHtyKEd2d1d0wHdxNDwC91Tx92Q25lz38G5kz1FnFjljDyTwjyaKjxtkLxT3KABXVhzxaK6NaF6Pz2FizytnNj5jzzqHqJKGL1Lx8PLxvzwKr04Bf0n2vGnybxbznztXWGxGAOmxRl737xxlW2HwRVHy1w4Anzfxn14B1yxgv1bHphX3QHX3Kh9w4G3yJHEH3xnjFCPwk0OG3hP0p3BQvyvyKxKzKyV3vyf1WJuM2M/22Lq1bD/woJ2JTn2Oa1IDAKeEoCgK8AaVgI0Emi8A0BpGN1WFWBpFnDiUwNvxhI4EIKFym2xKK1WF4BwP0iWLvyZIRm7xmB8iSHsEkCAA%3D%3D))